In [2]:
!pip install pandas

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.9/10.9 MB 12.7 MB/s  0:00:00m0:00:010:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 28.4 MB/s  0:00:00m0:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [pandas]2m1/2 [pandas]

[notice] A new release of pip is available: 26.0.1 -> 26.1.1
[notice] To update, run: pip install --upgrade pip


In [2]:
import pandas as pd 
import numpy as np

In [3]:
import ensurepip
ensurepip.bootstrap()

Looking in links: /tmp/tmp8j15oldr


In [4]:
import sys
!{sys.executable} -m pip install openpyxl


[notice] A new release of pip is available: 26.0.1 -> 26.1.1
[notice] To update, run: pip install --upgrade pip


In [7]:
df = pd.read_excel('Telco_customer_churn.xlsx')
df

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,2569-WGERO,1,United States,California,Landers,92285,"34.341737, -116.539416",34.341737,-116.539416,Female,...,Two year,Yes,Bank transfer (automatic),21.15,1419.4,No,0,45,5306,NaN
7039,6840-RESVB,1,United States,California,Adelanto,92301,"34.667815, -117.536183",34.667815,-117.536183,Male,...,One year,Yes,Mailed check,84.80,1990.5,No,0,59,2140,NaN
7040,2234-XADUH,1,United States,California,Amboy,92304,"34.559882, -115.637164",34.559882,-115.637164,Female,...,One year,Yes,Credit card (automatic),103.20,7362.9,No,0,71,5560,NaN
7041,4801-JZAZL,1,United States,California,Angelus Oaks,92305,"34.1678, -116.86433",34.167800,-116.864330,Female,...,Month-to-month,Yes,Electronic check,29.60,346.45,No,0,59,2793,NaN


In [8]:
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


In [9]:
df.shape

(7043, 33)

#### Cleaning data 

In [11]:
df.isnull().sum()

CustomerID              0
Count                   0
Country                 0
State                   0
City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges           0
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64

In [12]:
df.columns

Index(['CustomerID', 'Count', 'Country', 'State', 'City', 'Zip Code',
       'Lat Long', 'Latitude', 'Longitude', 'Gender', 'Senior Citizen',
       'Partner', 'Dependents', 'Tenure Months', 'Phone Service',
       'Multiple Lines', 'Internet Service', 'Online Security',
       'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV',
       'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method',
       'Monthly Charges', 'Total Charges', 'Churn Label', 'Churn Value',
       'Churn Score', 'CLTV', 'Churn Reason'],
      dtype='str')

In [13]:
df.drop_duplicates()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,2569-WGERO,1,United States,California,Landers,92285,"34.341737, -116.539416",34.341737,-116.539416,Female,...,Two year,Yes,Bank transfer (automatic),21.15,1419.4,No,0,45,5306,NaN
7039,6840-RESVB,1,United States,California,Adelanto,92301,"34.667815, -117.536183",34.667815,-117.536183,Male,...,One year,Yes,Mailed check,84.80,1990.5,No,0,59,2140,NaN
7040,2234-XADUH,1,United States,California,Amboy,92304,"34.559882, -115.637164",34.559882,-115.637164,Female,...,One year,Yes,Credit card (automatic),103.20,7362.9,No,0,71,5560,NaN
7041,4801-JZAZL,1,United States,California,Angelus Oaks,92305,"34.1678, -116.86433",34.167800,-116.864330,Female,...,Month-to-month,Yes,Electronic check,29.60,346.45,No,0,59,2793,NaN


In [14]:
df.shape

(7043, 33)

##### Here this line tells that , The columns Total charges looks like having all the numerics but sometimes pandas read it as string/text soo here 
- pd.to_numeric() funtion is used to convert all the values in the Total charges column to the numeric and,
- errors='coerce' this tells that is pandas finds the value which cannot be converted into the numbers then replace it with the NaN(missing value)

In [15]:
df["Total Charges"] = pd.to_numeric(df["Total Charges"] , errors='coerce')

##### Here we are filling the missing values using function fillna() and the median of the Total charges column 

In [16]:
df["Total Charges"] = df["Total Charges"].fillna(df["Total Charges"].median())

##### Here we are dividing the customers into 3 groups :
- 0-12 months - New customer (the customer who stayed for 12 months or less )
- 12-36 months - Medium customer (the customer who stayed more tham 12 months or less than 36 months)
- other - Long term customer (the customer who stayed more than 36 months)

In [17]:
def customer_type(tenure):
    if tenure <= 12:
        return "New Customer"
    elif tenure <= 36:
        return "Medium Customer"
    else:
        return "Long Term Customer"

df["Customer_Type"] = df["Tenure Months"].apply(customer_type)

##### Here we are filtering only the churned customers because the main aim of our project is to find out the reason why the customers are churned from our company , so we are taking only the churned customers columns 

In [18]:
churned_customers = df[df["Churn Label"] == "Yes"]

In [19]:
churned_customers

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason,Customer_Type
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer,New Customer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved,New Customer
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Yes,Electronic check,99.65,820.50,Yes,1,86,5372,Moved,New Customer
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved,Medium Customer
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1,89,5340,Competitor had better devices,Long Term Customer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1864,1699-HPSBG,1,United States,California,Palm Springs,92264,"33.745746, -116.514215",33.745746,-116.514215,Male,...,Yes,Electronic check,59.80,727.80,Yes,1,69,4581,Competitor offered higher download speeds,New Customer
1865,8775-CEBBJ,1,United States,California,Seeley,92273,"32.790282, -115.689559",32.790282,-115.689559,Female,...,Yes,Bank transfer (automatic),44.20,403.35,Yes,1,87,2029,Competitor offered higher download speeds,New Customer
1866,6894-LFHLY,1,United States,California,Vidal,92280,"34.264124, -114.717964",34.264124,-114.717964,Male,...,Yes,Electronic check,75.75,75.75,Yes,1,97,5571,Competitor offered higher download speeds,New Customer
1867,0639-TSIQW,1,United States,California,White Water,92282,"33.972293, -116.654195",33.972293,-116.654195,Female,...,Yes,Credit card (automatic),102.95,6886.25,Yes,1,70,5620,Competitor offered more data,Long Term Customer


#### Filtering

##### Now we need to filter our data by the contract of a customer because :
- Month-to-Month customer - this customer can leave easily more chances of a churn 
- one year contract - this customer has taken 1 year contract soo this customer should stay for 1 year atleast 
- two year contract - this customer has taken 2 years contract so the customer should stay for 2 years 

1. Filtering the data by Month-to-Month customers 

In [20]:
monthly_customers = df[df["Contract"] == "Month-to-month"]
monthly_customers

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason,Customer_Type
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer,New Customer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved,New Customer
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Yes,Electronic check,99.65,820.50,Yes,1,86,5372,Moved,New Customer
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved,Medium Customer
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1,89,5340,Competitor had better devices,Long Term Customer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7032,2235-DWLJU,1,United States,California,Thousand Palms,92276,"33.849263, -116.382778",33.849263,-116.382778,Female,...,Yes,Electronic check,44.40,263.05,No,0,71,4611,NaN,New Customer
7033,0871-OPBXW,1,United States,California,Twentynine Palms,92277,"34.17211, -115.769773",34.172110,-115.769773,Female,...,Yes,Mailed check,20.05,39.25,No,0,80,5191,NaN,New Customer
7035,9767-FFLEM,1,United States,California,Westmorland,92281,"33.03679, -115.60503",33.036790,-115.605030,Male,...,Yes,Credit card (automatic),69.50,2625.25,No,0,35,4591,NaN,Long Term Customer
7036,8456-QDAVC,1,United States,California,Winterhaven,92283,"32.852947, -114.850784",32.852947,-114.850784,Male,...,Yes,Bank transfer (automatic),78.70,1495.10,No,0,20,2464,NaN,Medium Customer


##### Here we are again filtering this by the customers who has taken the monthy contract how many number of customers has churned 

In [21]:
monthly_churned_customers = df[(df["Churn Label"] == "Yes") & (df["Contract"] == "Month-to-month")]
monthly_churned_customers

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason,Customer_Type
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer,New Customer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved,New Customer
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Yes,Electronic check,99.65,820.50,Yes,1,86,5372,Moved,New Customer
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved,Medium Customer
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1,89,5340,Competitor had better devices,Long Term Customer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1863,1122-JWTJW,1,United States,California,Ocotillo,92259,"32.698964, -115.886656",32.698964,-115.886656,Male,...,Yes,Mailed check,70.65,70.65,Yes,1,82,3463,Competitor had better devices,New Customer
1865,8775-CEBBJ,1,United States,California,Seeley,92273,"32.790282, -115.689559",32.790282,-115.689559,Female,...,Yes,Bank transfer (automatic),44.20,403.35,Yes,1,87,2029,Competitor offered higher download speeds,New Customer
1866,6894-LFHLY,1,United States,California,Vidal,92280,"34.264124, -114.717964",34.264124,-114.717964,Male,...,Yes,Electronic check,75.75,75.75,Yes,1,97,5571,Competitor offered higher download speeds,New Customer
1867,0639-TSIQW,1,United States,California,White Water,92282,"33.972293, -116.654195",33.972293,-116.654195,Female,...,Yes,Credit card (automatic),102.95,6886.25,Yes,1,70,5620,Competitor offered more data,Long Term Customer


##### Here is the final summary of the above 2 filterings :
- Total number of customers who has taken the monthly contract - 3875 customers 
- Total number of customers who has taken the monthly contract and churned the company - 1655 customers 
- Total number of customers who has taken the monthly contract but not churned the company - 2220 customers 

2. Filtering the customers who has taken the one year contract and again filtering them by how many of them has churned the company 

In [22]:
yearly_customer1 = df[df["Contract"] == "One year"]
yearly_customer1

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason,Customer_Type
41,0691-JVSYA,1,United States,California,Pasadena,91103,"34.167465, -118.165327",34.167465,-118.165327,Female,...,Yes,Bank transfer (automatic),94.85,5000.20,Yes,1,80,4328,Network reliability,Long Term Customer
80,2321-OMBXY,1,United States,California,Carlsbad,92009,"33.098018, -117.258203",33.098018,-117.258203,Female,...,No,Credit card (automatic),80.30,3058.65,Yes,1,81,2071,Limited range of services,Long Term Customer
82,5020-ZSTTY,1,United States,California,El Cajon,92020,"32.79697, -116.969082",32.796970,-116.969082,Female,...,No,Bank transfer (automatic),82.45,4350.10,Yes,1,83,4149,Long distance charges,Long Term Customer
86,4322-RCYMT,1,United States,California,Palomar Mountain,92060,"33.309852, -116.823091",33.309852,-116.823091,Male,...,Yes,Bank transfer (automatic),75.80,5293.95,Yes,1,81,4903,Service dissatisfaction,Long Term Customer
125,8205-OTCHB,1,United States,California,Riverside,92506,"33.930931, -117.361788",33.930931,-117.361788,Male,...,Yes,Bank transfer (automatic),43.75,903.60,Yes,1,66,3099,Network reliability,Medium Customer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7026,7203-OYKCT,1,United States,California,Palo Verde,92266,"33.3249, -114.758334",33.324900,-114.758334,Male,...,Yes,Electronic check,104.95,7544.30,No,0,20,5760,NaN,Long Term Customer
7034,3605-JISKB,1,United States,California,Twentynine Palms,92278,"34.457829, -116.139589",34.457829,-116.139589,Male,...,No,Credit card (automatic),60.00,3316.10,No,0,71,4212,NaN,Long Term Customer
7037,7750-EYXWZ,1,United States,California,Yucca Valley,92284,"34.159534, -116.425984",34.159534,-116.425984,Female,...,No,Electronic check,60.65,743.30,No,0,24,3740,NaN,New Customer
7039,6840-RESVB,1,United States,California,Adelanto,92301,"34.667815, -117.536183",34.667815,-117.536183,Male,...,Yes,Mailed check,84.80,1990.50,No,0,59,2140,NaN,Medium Customer


In [23]:
yearly_customer1_churned = df[(df["Contract"] == "One year") & (df["Churn Label"] == "Yes")]
yearly_customer1_churned

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason,Customer_Type
41,0691-JVSYA,1,United States,California,Pasadena,91103,"34.167465, -118.165327",34.167465,-118.165327,Female,...,Yes,Bank transfer (automatic),94.85,5000.20,Yes,1,80,4328,Network reliability,Long Term Customer
80,2321-OMBXY,1,United States,California,Carlsbad,92009,"33.098018, -117.258203",33.098018,-117.258203,Female,...,No,Credit card (automatic),80.30,3058.65,Yes,1,81,2071,Limited range of services,Long Term Customer
82,5020-ZSTTY,1,United States,California,El Cajon,92020,"32.79697, -116.969082",32.796970,-116.969082,Female,...,No,Bank transfer (automatic),82.45,4350.10,Yes,1,83,4149,Long distance charges,Long Term Customer
86,4322-RCYMT,1,United States,California,Palomar Mountain,92060,"33.309852, -116.823091",33.309852,-116.823091,Male,...,Yes,Bank transfer (automatic),75.80,5293.95,Yes,1,81,4903,Service dissatisfaction,Long Term Customer
125,8205-OTCHB,1,United States,California,Riverside,92506,"33.930931, -117.361788",33.930931,-117.361788,Male,...,Yes,Bank transfer (automatic),43.75,903.60,Yes,1,66,3099,Network reliability,Medium Customer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1820,5906-CVLHP,1,United States,California,Rowland Heights,91748,"33.976753, -117.897367",33.976753,-117.897367,Female,...,Yes,Credit card (automatic),54.15,2319.80,Yes,1,92,2761,Attitude of service provider,Long Term Customer
1833,8267-ZNYVZ,1,United States,California,Spring Valley,91977,"32.726627, -116.994608",32.726627,-116.994608,Female,...,Yes,Electronic check,103.75,3361.05,Yes,1,81,3824,Attitude of service provider,Medium Customer
1850,6664-FPDAC,1,United States,California,San Diego,92115,"32.762506, -117.07245",32.762506,-117.072450,Female,...,Yes,Electronic check,111.95,6418.90,Yes,1,91,4060,Attitude of service provider,Long Term Customer
1855,4174-LPGTI,1,United States,California,Beaumont,92223,"33.946982, -116.977672",33.946982,-116.977672,Female,...,Yes,Bank transfer (automatic),66.50,2728.60,Yes,1,65,2517,Moved,Long Term Customer


##### Finally the summary of this filtering is :
- Total number of customer taken one year contract - 1473 customers 
- Total number of customers churned by taking one year contract - 166 customers 
- Total number of customers who has taken one year contract but not churned the company - 1307 customers 

3. Filtering by the customers who has taken the 2 year contract and again filtering them by the chustomers who has churned the company 

In [24]:
yearly_customer2 = df[df["Contract"] == "Two year"]
yearly_customer2

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason,Customer_Type
24,3192-NQECA,1,United States,California,Santa Monica,90403,"34.031529, -118.491156",34.031529,-118.491156,Male,...,Yes,Bank transfer (automatic),110.00,7611.85,Yes,1,86,5034,Lack of affordable download/upload speed,Long Term Customer
65,6323-AYBRX,1,United States,California,Ontario,91762,"34.057256, -117.667677",34.057256,-117.667677,Male,...,No,Mailed check,19.35,1099.60,Yes,1,83,5084,Price too high,Long Term Customer
87,6680-NENYN,1,United States,California,Pauma Valley,92061,"33.313828, -116.940501",33.313828,-116.940501,Female,...,Yes,Electronic check,104.60,4759.85,Yes,1,91,5250,Lack of self-service on Website,Long Term Customer
202,4853-RULSV,1,United States,California,Miramonte,93641,"36.696759, -119.024051",36.696759,-119.024051,Male,...,Yes,Credit card (automatic),104.00,7250.15,Yes,1,80,4544,Attitude of service provider,Long Term Customer
248,2834-JRTUA,1,United States,California,Martinez,94553,"38.014457, -122.115432",38.014457,-122.115432,Male,...,Yes,Electronic check,108.05,7532.15,Yes,1,81,5200,Competitor had better devices,Long Term Customer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7023,4807-IZYOZ,1,United States,California,North Palm Springs,92258,"33.906496, -116.569499",33.906496,-116.569499,Female,...,No,Bank transfer (automatic),20.65,1020.75,No,0,70,4725,NaN,Long Term Customer
7024,9710-NJERN,1,United States,California,Palm Desert,92260,"33.694501, -116.412711",33.694501,-116.412711,Female,...,No,Mailed check,20.15,826.00,No,0,44,2007,NaN,Long Term Customer
7031,9281-CEDRU,1,United States,California,Salton City,92275,"33.28156, -115.955541",33.281560,-115.955541,Female,...,No,Bank transfer (automatic),64.10,4326.25,No,0,48,5553,NaN,Long Term Customer
7038,2569-WGERO,1,United States,California,Landers,92285,"34.341737, -116.539416",34.341737,-116.539416,Female,...,Yes,Bank transfer (automatic),21.15,1419.40,No,0,45,5306,NaN,Long Term Customer


In [25]:
yearly_customer2_churned = df[(df["Contract"] == "Two year") & (df["Churn Label"] == "Yes")]
yearly_customer2_churned

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason,Customer_Type
24,3192-NQECA,1,United States,California,Santa Monica,90403,"34.031529, -118.491156",34.031529,-118.491156,Male,...,Yes,Bank transfer (automatic),110.00,7611.85,Yes,1,86,5034,Lack of affordable download/upload speed,Long Term Customer
65,6323-AYBRX,1,United States,California,Ontario,91762,"34.057256, -117.667677",34.057256,-117.667677,Male,...,No,Mailed check,19.35,1099.60,Yes,1,83,5084,Price too high,Long Term Customer
87,6680-NENYN,1,United States,California,Pauma Valley,92061,"33.313828, -116.940501",33.313828,-116.940501,Female,...,Yes,Electronic check,104.60,4759.85,Yes,1,91,5250,Lack of self-service on Website,Long Term Customer
202,4853-RULSV,1,United States,California,Miramonte,93641,"36.696759, -119.024051",36.696759,-119.024051,Male,...,Yes,Credit card (automatic),104.00,7250.15,Yes,1,80,4544,Attitude of service provider,Long Term Customer
248,2834-JRTUA,1,United States,California,Martinez,94553,"38.014457, -122.115432",38.014457,-122.115432,Male,...,Yes,Electronic check,108.05,7532.15,Yes,1,81,5200,Competitor had better devices,Long Term Customer
480,6990-YNRIO,1,United States,California,Agoura Hills,91301,"34.129058, -118.759788",34.129058,-118.759788,Male,...,No,Bank transfer (automatic),108.65,6937.95,Yes,1,79,6120,Attitude of support person,Long Term Customer
526,1447-PJGGA,1,United States,California,San Diego,92101,"32.72523, -117.171346",32.725230,-117.171346,Female,...,Yes,Electronic check,95.25,5464.65,Yes,1,93,5789,Competitor offered higher download speeds,Long Term Customer
604,5440-FLBQG,1,United States,California,Santa Barbara,93110,"34.437945, -119.77191",34.437945,-119.771910,Male,...,Yes,Bank transfer (automatic),108.40,7318.20,Yes,1,73,4708,Attitude of service provider,Long Term Customer
629,5236-PERKL,1,United States,California,Pearblossom,93553,"34.445239, -117.894868",34.445239,-117.894868,Female,...,Yes,Bank transfer (automatic),112.95,6465.00,Yes,1,89,4738,Competitor offered higher download speeds,Long Term Customer
729,5013-SBUIH,1,United States,California,Stockton,95202,"37.959706, -121.287669",37.959706,-121.287669,Female,...,Yes,Electronic check,109.65,5551.15,Yes,1,84,5855,Competitor had better devices,Long Term Customer


##### Finally the summary of this filtering is :
- Total number of customers who has taken two years contract - 1695 customers 
- Total number of customers who has churned this company - 48 customers 
- Total number of customers who has taken 2 years contract but not churned the company - 1647 customers 

### Grouping 

##### Now we need to know in total how many customers left the company and how many customers stayed in the company , we are grouping the customers into 2 groups left customers and stayed customers 

In [26]:
churn_count = df["Churn Label"].value_counts()
churn_count

Churn Label
No     5174
Yes    1869
Name: count, dtype: int64

##### Here this result tells us that :
- Total customers churned - 1869
- Total chustomers stayed - 5174

##### Now lets us check what makes the customers churn whether the higher monthly charges made the customers churn lets see 

In [27]:
avg_charges = df.groupby("Churn Label")["Monthly Charges"].mean()
avg_charges

Churn Label
No     61.265124
Yes    74.441332
Name: Monthly Charges, dtype: float64

##### so here this result tells us that the percentage of the churned customer based on the Monthly charges 
- churned cutomers : 74.441332
- stayed customers / Not churned customers : 61.265124

##### Now let us group the churned and stayed cutomers by the Contract 

In [28]:
contract_churn = df.groupby(["Contract", "Churn Label"]).size()

print(contract_churn)

Contract        Churn Label
Month-to-month  No             2220
                Yes            1655
One year        No             1307
                Yes             166
Two year        No             1647
                Yes              48
dtype: int64


#### Pivot Tables 

##### 1. Churn by Contract 

In [29]:
pivot_contract = pd.pivot_table(
    df,
    index="Contract",
    columns="Churn Label",
    values="CustomerID",
    aggfunc="count"
)

print(pivot_contract)

Churn Label       No   Yes
Contract                  
Month-to-month  2220  1655
One year        1307   166
Two year        1647    48


##### 2. Churn by Payment Method

In [30]:
pivot_payment = pd.pivot_table(
    df,
    index="Payment Method",
    columns="Churn Label",
    values="CustomerID",
    aggfunc="count"
)

print(pivot_payment)

Churn Label                  No   Yes
Payment Method                       
Bank transfer (automatic)  1286   258
Credit card (automatic)    1290   232
Electronic check           1294  1071
Mailed check               1304   308


##### 3. Average Monthly charges by Contract

In [31]:
pivot_avg_charges = pd.pivot_table(
    df,
    index="Contract",
    values="Monthly Charges",
    aggfunc="mean"
)

print(pivot_avg_charges)

                Monthly Charges
Contract                       
Month-to-month        66.398490
One year              65.048608
Two year              60.770413


##### Saving all the excel work into the excel file 

In [32]:
df.to_excel("cleaned_customer_churn.xlsx", index=False)

In [33]:
with pd.ExcelWriter("customer_churn_excel_analysis.xlsx", engine="openpyxl") as writer:
    df.to_excel(writer, sheet_name="Cleaned_Data", index=False)
    pivot_contract.to_excel(writer, sheet_name="Churn_by_Contract")
    pivot_payment.to_excel(writer, sheet_name="Churn_by_Payment")
    pivot_avg_charges.to_excel(writer, sheet_name="Avg_Charges")